cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java

```diff
@@ -515,6 +518,37 @@ Collection<WorkflowTable.WorkflowRunRecord> retrieveSpacedRecords(WorkflowId wor
   List<ProgramHistory> getRuns(Collection<ProgramId> programs, ProgramRunStatus status,
                                long startTime, long endTime, int limitPerProgram);
 
+
+  /**
+   * Get application state.
+   *
+   * @param request a {@link AppStateKey} object.
+   * @return state of application
+   */
+  Optional<byte[]> getState(AppStateKey request);
```

Comment:
all the methods should throw ApplicationNotFoundException.

As written, I think the endpoints should be returning 500 errors if the app doesn't exist instead of the desired 404. Can you add a unit test to make sure this case is covered?